File: src/main/java/gargoyle/xenox/game/scr/game/game/Level.java
package gargoyle.xenox.game.scr.game.game;

import gargoyle.xenox.Xenox;
import gargoyle.xenox.util.ini.PropertyMap;
import gargoyle.xenox.util.log.Log;
import gargoyle.xenox.util.res.Res;

import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.net.URL;
import java.nio.charset.Charset;

public class Level {
    private static final int DEFAULT_BALLS = 80;
    private static final int DEFAULT_PERCENT = 80;
    private static final int DEFAULT_SPEED = 50;
    private static final String PARAM_BALLS = "balls";
    private static final String PARAM_COVER = "cover";
    private static final String PARAM_HEIGHT = "height";
    private static final String PARAM_IMAGE = "image";
    private static final String PARAM_LIVES = "lives";
    private static final String PARAM_MUSIC = "music";
    private static final String PARAM_PERCENT = "percent";
    private static final String PARAM_SPEED = "speed";
    private static final String PARAM_WIDTH = "width";
    private static final String PROP_TITLE = "title";
    private String title;
    private int balls = DEFAULT_BALLS;
    private URL cover;
    private int height = Xenox.PF_HEIGHT;
    private URL image;
    private int lives = Xenox.LIVES;
    private URL music;
    private int percent = DEFAULT_PERCENT;
    private int speed = DEFAULT_SPEED;
    private int width = Xenox.PF_WIDTH;

    public Level(URL url, Charset charset) {
        try (Reader reader = new InputStreamReader(url.openStream(), charset)) {
            PropertyMap file = new PropertyMap(reader);
            title = file.get(PROP_TITLE);
            balls = file.get(PARAM_BALLS, DEFAULT_BALLS);
            height = file.get(PARAM_HEIGHT, Xenox.PF_HEIGHT);
            lives = file.get(PARAM_LIVES, Xenox.LIVES);
            percent = file.get(PARAM_PERCENT, DEFAULT_PERCENT);
            speed = file.get(PARAM_SPEED, DEFAULT_SPEED);
            width = file.get(PARAM_WIDTH, Xenox.PF_WIDTH);
            cover = Res.url(url, file.get(PARAM_COVER));
            image = Res.url(url, file.get(PARAM_IMAGE));
            music = Res.url(url, file.get(PARAM_MUSIC));
        } catch (IOException e) {
            Log.error(String.format("cannot load level from %s", url), e);
        }
    }

    public int getBalls() {
        return balls;
    }

    public URL getCover() {
        return cover;
    }

    public int getHeight() {
        return height;
    }

    public URL getImage() {
        return image;
    }

    public int getLives() {
        return lives;
    }

    public URL getMusic() {
        return music;
    }

    public int getPercent() {
        return percent;
    }

    public int getSpeed() {
        return speed;
    }

    public String getTitle() {
        return title;
    }

    public int getWidth() {
        return width;
    }
}

File: src/main/kotlin/gargoyle/xenox/game/scr/game/game/Level.kt
package gargoyle.xenox.game.scr.game.game

import gargoyle.xenox.Xenox
import gargoyle.xenox.util.ini.PropertyMap
import gargoyle.xenox.util.log.Log
import gargoyle.xenox.util.res.Res
import java.io.IOException
import java.io.InputStreamReader
import java.net.URL
import java.nio.charset.Charset

class Level(url: URL, charset: Charset) {
    var title: String? = null
    var balls = DEFAULT_BALLS
    var cover: URL? = null
    var height: Int = Xenox.PF_HEIGHT
    var image: URL? = null
    var lives: Int = Xenox.LIVES
    var music: URL? = null
    var percent = DEFAULT_PERCENT
    var speed = DEFAULT_SPEED
    var width: Int = Xenox.PF_WIDTH

    init {
        try {
            InputStreamReader(url.openStream(), charset).use { reader ->
                val file = PropertyMap(reader)
                title = file[PROP_TITLE]
                balls = file[PARAM_BALLS, DEFAULT_BALLS]
                height = file[PARAM_HEIGHT, Xenox.PF_HEIGHT]
                lives = file[PARAM_LIVES, Xenox.LIVES]
                percent = file[PARAM_PERCENT, DEFAULT_PERCENT]
                speed = file[PARAM_SPEED, DEFAULT_SPEED]
                width = file[PARAM_WIDTH, Xenox.PF_WIDTH]
                cover = Res.url(url, file[PARAM_COVER]!!)
                image = Res.url(url, file[PARAM_IMAGE]!!)
                music = Res.url(url, file[PARAM_MUSIC]!!)
            }
        } catch (e: IOException) {
            Log.error(String.format("cannot load level from %s", url), e)
        }
    }

    companion object {
        private const val DEFAULT_BALLS = 80
        private const val DEFAULT_PERCENT = 80
        private const val DEFAULT_SPEED = 50
        private const val PARAM_BALLS = "balls"
        private const val PARAM_COVER = "cover"
        private const val PARAM_HEIGHT = "height"
        private const val PARAM_IMAGE = "image"
        private const val PARAM_LIVES = "lives"
        private const val PARAM_MUSIC = "music"
        private const val PARAM_PERCENT = "percent"
        private const val PARAM_SPEED = "speed"
        private const val PARAM_WIDTH = "width"
        private const val PROP_TITLE = "title"
    }
}
